rancher/rancher
# Cluster agent backend logic

## Issue: 


https://github.com/rancher/rancher/issues/41035
https://github.com/rancher/rancher/issues/41132 (update RKE2 cluster agent manifest logic, may block #41035)
 
## Problem


Add backend logic for Cluster and Fleet Agent Customization.
 
## Solution


Add backend logic to use values passed in ClusterAgentDeploymentCustomization and FleetAgentDeploymentCustomization in a separate PR, target merge of 4/26/2023. Includes:

- [x] Updating YAML manifest generation for cluster-agent deployment to use values passed
- [x] Updating cluster agent redeploy
- [x] Sync of these `ClusterAgentDeploymentCustomization` and `FleetAgentDeploymentCustomization` fields between management and provisioning cluster objects (also includes using cluster-agent-default-affinity setting for affinity)
- [x] Application of values from `FleetAgentDeploymentCustomization` field from provisioning cluster object to Fleet cluster object (also includes using fleet-agent-default-affinity setting for affinity)
- [x] Add fields to etcd snapshot restore [example](https://github.com/rancher/rancher/pull/40611/files)
 
Notes

It was overlooked that to update cluster agent redeploy, we need to have the agent customizations added to the v3 cluster Status so the values can be compared to determine whether we need to redeploy the agent or not. As such, `AppliedClusterAgentDeploymentCustomization` and `AppliedFleetAgentDeploymentCustomization` structs have been added to the Status.

From the design plan for this feature, we want a bi-directional sync for cluster agent and fleet fields between the mgmt and provisioning cluster objects and a uni-directional sync for fleet fields from the v1 provisioning object to the fleet cluster object. I have made sure to sync the fleet fields back to the v3 mgmt object so the user can view all agent customization in the mgmt cluster Spec in the UI.

4/28 I did see that passing in default `cluster-agent-default-affinity` from settings to the agent manifest instead of hardcoding it on the template reordered the affinity rules but we have no control over that and I did not see any incorrect agent redeploys as a result.

## Testing


## Engineering Testing
### Manual Testing


Smoke tested on Rancher v2.7-head (2.7.2)
- [x]  RKE2 EC2 single node cluster with no cluster-agent customization, default values
- [x]  RKE2 EC2 single node cluster with user defined fields via **Edit Yaml (edit cluster object in UI)**
- [x]  RKE EC2 single node cluster, default values
- [x]  RKE EC2 single node cluster, user defined values
- [x] EKS 2 node cluster, user defined values
- [x] Generic custom cluster (run registration cmd on an EC2 node to import it), user defined values - _PENDING_
- [x]  Upgrade 2.7-head -> branch with updates, verify that the planner does not reconcile/deliver a new plan/there are no changes to agent yaml _VERIFIED_

For each cluster type, verify
- cluster comes up active
- v1 cluster and v3 mgmt objects have no/the correct agent customization defined (`v1.Spec`, `v3.Spec`, `v3.Status. AppliedClusterAgentDeploymentCustomization`)
- cluster/fleet agent pod yaml has correct defaults/user defined values
- updating a user defined value redeploys the agent

<details>

<summary>Test yaml for cluster agent (since UI is not available yet)</summary>

```
clusterAgentDeploymentCustomization:
  appendTolerations:
    - effect: NoSchedule
      key: node-role.kubernetes.io/controlplane
      value: 'true'
  overrideAffinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - preference:
            matchExpressions:
              - key: cattle.io/cluster-agent
                operator: In
                values:
                  - 'true'
          weight: 1
  overrideResourceRequirements:
    limits:
      cpu: '0'
      memory: '0'
    requests:
      cpu: '0'
      memory: '0'
```

</details>

<details>

<summary>Test yaml for fleet agent</summary>

```
fleetAgentDeploymentCustomization:
  appendTolerations:
    - effect: NoSchedule
      key: node-role.kubernetes.io/controlplane
      value: 'true'
  overrideAffinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - preference:
            matchExpressions:
              - key: cattle.io/fleet-agent
                operator: In
                values:
                  - 'true'
          weight: 1
  overrideResourceRequirements:
    limits:
      cpu: '0'
      memory: '0'
    requests:
      cpu: '0'
      memory: '0'
```

</details>

Additional test
- [x] Edit rke2 cluster to have invalid tolerations, affinity, and resources. Verify `Error while applying agent YAML` is bubbled up instead of attempting an agent redeploy

### Automated Testing


@Oats87 has added a unit test in `import_test.go` that compares cluster agent manifest files post-upgrade to make sure they are the same.

## QA Testing Considerations

 
### Regressions Considerations
